draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -379,23 +380,41 @@ Algorithms in this document make use of utility functions described below.
 - round(x): floor(ceil(2\*x)/2), i.e. rounding to the nearest
   integer, rounding down when the remainder is precisely 0.5.
 
-- HashToBase(H,x): H(x)[0:round(log2(p))] mod p.
-  This method is parametrized by the cryptographic hash function H, which
-  outputs at least round(log2(p)) bits, where p is the prime order of base
-  field Fp. The function first computes the bitstring H(x), and truncates this
-  down to round(log2(p)) bits. The result of this is converted to an integer,
-  and reduced modulo p to give an element of Fp. We suggest a more detailed
-  algorithm in {{hashtobase}}.
+- HashToBase(x, i).
+  This method is parametrized by p and H, where p is the prime order of
+  the base field Fp, and H is a cryptographic hash function which
+  outputs at least round(log2(p)) bits.
+  The function first hashes x, truncates the result, and converts to an integer,
+  and reduced modulo p to give an element of Fp. The function returns
+  this element of Fp, along with an additional bit, which can be used for
+  choosing square roots.
```

Comment:
Keeping one bit during all the mapping can result costly, if constant-time implementation is required.
A better approach is that every suite decides a criteria for selecting square roots. 
For example, for p=3 mod 4, one can choose the principal square root. 
And, for p=5 mod 8, one can choose the root x such that 0 <= x <= (p-1)/2. 
These criteria must be set by the suite.

Reply:
So I'm with you on the principal square root computation. But the idea of returning an additional bit from HashToBase is to be able to cover more elements in the encoding phase. Given a properly specified principal `sqrt` function, it is still useful to be able to do `CNEG(sqrt(x), s)`, making use of the additional bit of entropy. Or perhaps a better way would be to move it to the final computation. I.e. replace `return P` with `return CNEG(P, s)`.

I'm not sure I follow the claim about making it costly and relevance to constant time operations though.